File: python-backend/backend-python/core/eventBus.py
# backend-python/core/eventBus.py
import asyncio
from collections import defaultdict
from utils.logger import logger

class EventBus:
    def __init__(self):
        self.subscribers = defaultdict(list)

    def subscribe(self, event_type: str, callback):
        self.subscribers[event_type].append(callback)
        logger.log(f"[EventBus] Subscribed to event: {event_type}")

    def publish(self, event_type: str, data: dict):
        logger.log(f"[EventBus] Publishing event: {event_type} → {data.get('collection', '')}:{data.get('key', '')}")
        for callback in self.subscribers[event_type]:
            # Call asynchronously if coroutine
            if asyncio.iscoroutinefunction(callback):
                asyncio.create_task(callback(data))
            else:
                callback(data)

eventBus = EventBus()

File: typescript-backend/lib/core/eventBus.ts
// src/core/eventBus.ts
export const eventBus: Record<string, Function[]> = {};

export function subscribe(channel: string, callback: Function) {
  if (!eventBus[channel]) eventBus[channel] = [];
  eventBus[channel].push(callback);
}

export function publish(channel: string, data: any) {
  const subscribers = eventBus[channel] || [];
  subscribers.forEach(cb => cb(data));
}
